mu4/domain/notation/internal/notationplayback.cpp

```diff
@@ -0,0 +1,315 @@
+    default: {
+        LOGE() << "unknown midi type: " << type;
+    }
+    }
+    return ME_INVALID;
+}
+
+NotationPlayback::NotationPlayback(IGetScore* getScore)
+    : m_getScore(getScore)
+{
+}
+
+std::shared_ptr<MidiStream> NotationPlayback::midiStream() const
+{
+    Ms::Score* score = m_getScore->score();
+    if (!score) {
+        return nullptr;
+    }
+
+    std::shared_ptr<MidiStream> stream = std::make_shared<MidiStream>();
+
+    makeInitData(stream->initData, score);
+
+    stream->request.onReceive(this, [stream](uint32_t tick) {
+        UNUSED(tick);
+        stream->stream.close();
+    });
+
+    return stream;
+}
+
+void NotationPlayback::makeInitData(MidiData& data, Ms::Score* score) const
+{
+    MetaInfo meta;
+    makeMetaInfo(meta, score);
+
+    data.division = Ms::MScore::division;
+    data.tracks.resize(meta.tracksCount);
+
+    Ms::EventMap eventMap;
+    makeEventMap(eventMap, score);
+    fillTracks(data.tracks, eventMap, meta);
+
+    fillTempoMap(data.tempomap, score);
+
+    //fillMetronome(stream->initData.metronome, score, midiSpec);
+}
+
+void NotationPlayback::makeEventMap(Ms::EventMap& eventMap, Ms::Score* score) const
+{
+//    int unrenderedUtick = renderEventsStatus.occupiedRangeEnd(utick);
+//    while (unrenderedUtick - utick < minUtickBufferSize) {
+//        const MidiRenderer::Chunk chunk = midi.getChunkAt(unrenderedUtick);
+//        if (!chunk) {
+//            break;
+//        }
+//        renderChunk(chunk, &events);
+//        unrenderedUtick = renderEventsStatus.occupiedRangeEnd(utick);
+//    }
+
+    score->masterScore()->setExpandRepeats(true);
+    score->renderMidi(&eventMap, Ms::SynthesizerState());
+}
+
+void NotationPlayback::makeMetaInfo(MetaInfo& meta, const Ms::Score* score) const
+{
+    auto parts = score->parts();
+
+    meta.tracksCount = parts.size();
+
+    auto bankForInstrument = [](const Ms::Instrument* instr) {
+                                 //! NOTE Temporary solution
+                                 if (instr->useDrumset()) {
+                                     return 128;
+                                 }
+                                 return 0;
+                             };
+
+    for (int pi = 0; pi < parts.size(); ++pi) {
+        const Ms::Part* part = parts.at(pi);
+
+        const Ms::InstrumentList* instList = part->instruments();
+        for (auto it = instList->cbegin(); it != instList->cend(); ++it) {
+            const Ms::Instrument* instrument = it->second;
+
+            uint16_t bank = bankForInstrument(instrument);
+
+            for (const Ms::Channel* ch : instrument->channel()) {
+                ChanInfo chi;
+                chi.trackIdx = pi;
+                chi.bank = bank;
+                chi.program = ch->program();
+
+                meta.channels.insert({ ch->channel(), chi });
+            }
+        }
+    }
+}
+
+void NotationPlayback::fillTracks(std::vector<audio::midi::Track>& tracks, const Ms::EventMap& eventMap,
```

Comment:
Looked, I have no idea what makes sense to separate. It may be `findOrAddChannel`, but it is designed as a separate lambda function.

Everything seems simple.
We iterate over events
We find a track for each event.
Find or create a channel in a track
We put events in this channel
We are skipping some events that we do not need (at the moment)